src/github.com/mozilla-services/pushgo/simplepush/dynamo_store.go

```diff
@@ -0,0 +1,507 @@
+// errors. Safe to call multiple times. Implements Store.Close().
+func (s *DynamoDBStore) Close() (err error) {
+	return
+}
+
+// KeyToIDs extracts the hex-encoded device and channel IDs from a user-
+// readable primary key. Implements Store.KeyToIDs().
+func (*DynamoDBStore) KeyToIDs(key string) (uaid, chid string, err error) {
+	items := strings.SplitN(key, ".", 2)
+	if len(items) < 2 {
+		return "", "", nil
+	}
+	return items[0], items[1], nil
+}
+
+// IDsToKey generates a user-readable primary key from a (device ID, channel
+// ID) tuple. The primary key is encoded in the push endpoint URI. Implements
+// Store.IDsToKey().
+func (*DynamoDBStore) IDsToKey(uaid, chid string) (string, error) {
+	if len(uaid) == 0 || len(chid) == 0 {
+		return "", nil
+	}
+	return fmt.Sprintf("%s.%s", uaid, chid), nil
+}
+
+// Status queries whether memcached is available for reading and writing.
+// Implements Store.Status().
+func (s *DynamoDBStore) Status() (success bool, err error) {
+	success = false
+	tables, err := s.server.ListTables()
+	success = err == nil && s.containsTable(tables, s.tablename)
+	return true, nil
+}
+
+// Exists returns a Boolean indicating whether a device has previously
+// registered with the Simple Push server. Implements Store.Exists().
+func (s *DynamoDBStore) Exists(uaid string) bool {
+	res, err := s.table.CountQuery([]dynamodb.AttributeComparison{
+		*dynamodb.NewEqualStringAttributeComparison(UAID_LABEL, uaid),
+		*dynamodb.NewEqualStringAttributeComparison(CHID_LABEL, " "),
+	})
+	if err != nil {
+		if s.logger.ShouldLog(ERROR) {
+			s.logger.Error(DB_LABEL, "Exists failed",
+				LogFields{"error": err.Error()})
+		}
+		return false
+	}
+	return res > 0
+}
+
+// Register creates and stores a channel record for the given device ID and
+// channel ID. If version > 0, the record will be marked as active. Implements
+// Store.Register().
+func (s *DynamoDBStore) Register(uaid, chid string, version int64) (err error) {
+	// try to put the master record
+	now := s.getNow(0)
+	vers := strconv.FormatInt(version, 10)
+	success, err := s.table.PutItem(uaid, " ", []dynamodb.Attribute{
+		*dynamodb.NewNumericAttribute(MODD_LABEL, now),
+	})
+	if err != nil {
+		return
+	}
+	// now add the chid.
+	success, err = s.table.PutItem(uaid, chid, []dynamodb.Attribute{
+		*dynamodb.NewNumericAttribute(VERS_LABEL, vers),
+		*dynamodb.NewNumericAttribute(MODD_LABEL, now),
+	})
+	if err != nil {
+		return
+	}
+	if success != true {
+		err = ErrDynamoDBFailure
+	}
+	return
+}
+
+// Update updates the version for the given device ID and channel ID.
+// Implements Store.Update().
+func (s *DynamoDBStore) Update(uaid, chid string, version int64) (err error) {
+	now := s.getNow(0)
+	success, err := s.table.UpdateAttributes(&dynamodb.Key{uaid, chid},
+		[]dynamodb.Attribute{
+			*dynamodb.NewNumericAttribute(VERS_LABEL,
+				strconv.FormatInt(version, 10)),
+			*dynamodb.NewNumericAttribute(MODD_LABEL, now),
+		})
```

Comment:
Should this use a conditional to avoid updating it if another update sent closely wants an even higher version number? (ie, don't update unless this version number will be the highest)